src/EditorFeatures/Core/LanguageServer/Handlers/Completion/CompletionResolveHandler.cs

```diff
@@ -126,20 +126,28 @@ private static bool MatchesLSPCompletionItem(LSP.CompletionItem lspCompletionIte
             return string.Equals(originalDisplayText, completionItem.DisplayText);
         }
 
-        // Internal for testing
-        internal static async Task<LSP.TextEdit> GenerateTextEditAsync(
+        internal static async Task AddTextEditAsync(
+            LSP.CompletionItem lspItem,
             Document document,
+            SourceText documentText,
             CompletionService completionService,
             CompletionItem selectedItem,
             bool snippetsSupported,
+            TextSpan listSpan,
+            TextSpan? itemDefaultSpan,
             CancellationToken cancellationToken)
         {
-            var documentText = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
-
             var completionChange = await completionService.GetChangeAsync(
                 document, selectedItem, cancellationToken: cancellationToken).ConfigureAwait(false);
-            var completionChangeSpan = completionChange.TextChange.Span;
-            var newText = completionChange.TextChange.NewText;
+
+            // Use CompletionChange.TextChanges so that we can get minimal edits around the cursor for better filtering.
+            // For the rest of the edits that are not around the cursor, classify them as additional edits.
+            var mainEdit = completionChange.TextChanges.Single(change => change.Span.IntersectsWith(listSpan));
```

Comment:
An important thing to note for this edit is that it _must_ be replacing a single line (https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#completionItem):

```tx
 	/**
	 * An edit which is applied to a document when selecting this completion.
	 * When an edit is provided the value of `insertText` is ignored.
	 *
	 * *Note:* The range of the edit must be a single line range and it must
	 * contain the position at which completion has been requested.
```

If this is not a single line (and trust me, it's currently not guaranteed to be), you'll need to break that change up.

This also creates complications with filtering, as there is no separate range for what LSP clients should filter on. I've forwarded an email discussing the recent problem we fixed with async completion, where the prefix of the line was getting included in the change, but this is a general problem that will need to be addressed. The range of the `textEdit` is used as the range for filtering as well. I would bet VS works differently here, but in general the LSP spec will cause many completions to be immediately filtered out. I would recommend looking through the omnisharp-roslyn code [here](https://github.com/OmniSharp/omnisharp-roslyn/blob/master/src/OmniSharp.Roslyn.CSharp/Services/Completion/CompletionListBuilder_Sync.cs#L166-L209), where I've tried to document examples of what will happen for various scenarios.

Reply:
> If this is not a single line (and trust me, it's currently not guaranteed to be), you'll need to break that change up.

Good callout - currently VS doesn't care at all if the text edit spans multiple lines 😆.  And at least for override completion it doesn't impact filtering since VS has the behavior where we only resolve on commit for [complex textedits](https://sourceroslyn.io/#Microsoft.CodeAnalysis.Features/Completion/CompletionItem.cs,100).  So the VS client infers the range and seems to be filtering _ok_ for the most part, though I'll need to verify some of the scenarios you linked.  I'm hoping that they are covered by `IsComplexTextEdit`. 

I definitely do see the benefit of an explicit filter range though - instead of the client inferring the range for complex edits, we can just tell them what the filter range is without providing a text edit.